cranelift/codegen/src/ir/instructions.rs

```diff
@@ -42,12 +42,26 @@ pub struct BlockWithArgs {
 }
 
 impl BlockWithArgs {
-    /// TODO: comment about the layout of BlockWithArgs
+    // NOTE: We don't expose either `value_to_block` or `block_to_value` outside of this module
+    // because this operation is not generally safe. However, as the two share the same underlying
+    // layout, they can be stored in the same value pool.
+    //
+    // `BlockWithArgs` makes use of this shared layout by storing all of its contents (a block and
+    // its argument) in a single `EntityList`. This is a bit better than introducing a new entity
+    // type for the pair of a block name and the arguments entity list, as we don't pay any
+    // indirection penalty to get to the argument values -- they're stored in-line with the block
+    // in the same list.
+    //
+    // The `BlockWithArgs::new` function guarantees this layout by requiring a block argument
+    // that's written in as the first element of the EntityList. Any subsequent entries are always
+    // assumed to be real Values.
```

Comment:
This description of the `BlockWithArgs` invariants is exactly what I wanted to see. Perhaps most of it should be merged into the doc comment for the `BlockWithArgs::values` field above, though?

Reply:
Happy to move it!